index.js

```diff
@@ -16,41 +13,44 @@ var typedArrays = {
 	Uint16Array: true,
 	Uint32Array: true
 };
-
 var slice = String.prototype.slice;
-var toStrTags = {};
+var getpr = Object.getPrototypeOf;
 var gOPD = Object.getOwnPropertyDescriptor;
-if (hasToStringTag && gOPD && Object.getPrototypeOf) {
-	forEach(typedArrays, function (_, typedArray) {
-		var arr = new global[typedArray]();
-		if (!(Symbol.toStringTag in arr)) {
-			throw new EvalError('this engine has support for Symbol.toStringTag, but ' + typedArray + ' does not have the property! Please report this.');
+var isTypedArray;
+if (getpr && gOPD) {
+	if (typeof Symbol === 'function') {
+		if (typeof Symbol.toStringTag === 'symbol') {
+			isTypedArray = function (value) {
+				var typedArray, arr, proto, descriptor;
+				/* jscs:disable disallowNodeTypes */
+				for (typedArray in typedArrays) {
+					if (has(typedArrays, typedArray)) {
+						arr = new global[typedArray]();
+						if (!(Symbol.toStringTag in arr)) {
+							throw new EvalError('this engine has support for Symbol.toStringTag, but ' + typedArray + ' does not have the property! Please report this.');
+						}
+						proto = getpr(arr);
+						descriptor = gOPD(proto, Symbol.toStringTag);
+						if (!descriptor) {
+							proto = getpr(proto);
+							descriptor = gOPD(proto, Symbol.toStringTag);
+						}
+						if (descriptor) {
+							descriptor = typedArrays[descriptor.get.call(Object(value))];
+							if (descriptor) { return descriptor; }
+						}
+					}
+				}
+				/* jscs:enable disallowNodeTypes */
+				return false;
+			};
 		}
-		var proto = Object.getPrototypeOf(arr);
-		var descriptor = gOPD(proto, Symbol.toStringTag);
-		if (!descriptor) {
-			var superProto = Object.getPrototypeOf(proto);
-			descriptor = gOPD(superProto, Symbol.toStringTag);
-		}
-		toStrTags[typedArray] = descriptor.get;
-	});
+	}
+}
+if (!isTypedArray) {
+	isTypedArray = function (value) {
+		return typedArrays[slice.call(toStr.call(Object(value)), 8, -1)] || false;
+	};
 }
 
-var tryTypedArrays = function tryTypedArrays(value) {
-	var anyTrue = false;
-	forEach(toStrTags, function (getter, typedArray) {
-		if (!anyTrue) {
-			try {
-				anyTrue = getter.call(value) === typedArray;
-			} catch (e) { /**/ }
-		}
-	});
-	return anyTrue;
-};
-
-module.exports = function isTypedArray(value) {
```

Comment:
I intentionally return the same function in all cases - since the purpose of this PR is the "for-each" optimization, please revert all unrelated changes.